genie-agent/src/main/java/com/netflix/genie/agent/execution/services/DownloadServiceImpl.java

```diff
@@ -0,0 +1,100 @@
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services;
+
+import com.google.common.io.Files;
+import com.netflix.genie.agent.execution.exceptions.DownloadException;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Map;
+
+/**
+ * Cache-backed download service to retrieve a set of files and place them at an expected location, according
+ * to a given manifest.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+class DownloadServiceImpl implements DownloadService {
+    private final FetchingCacheService fetchingCacheService;
+
+    DownloadServiceImpl(final FetchingCacheService fetchingCacheService) {
+        this.fetchingCacheService = fetchingCacheService;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public DownloadManifest.Builder newManifestBuilder() {
+        return new DownloadManifestImpl.Builder();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void download(final DownloadManifest downloadsManifest) throws DownloadException {
+
+        // Validate all destination directories exist
+        for (final File targetDirectory : downloadsManifest.getTargetDirectories()) {
+            if (!targetDirectory.exists()) {
+                throw new DownloadException(
+                    "Target directory does not exist: " + targetDirectory.getAbsolutePath()
+                );
+            } else if (!targetDirectory.isDirectory()) {
+                throw new DownloadException(
+                    "Target directory is not a directory: " + targetDirectory.getAbsolutePath()
+                );
+            }
+        }
+
+        // Validate all target files don't exist
+        for (final File targetFile : downloadsManifest.getTargetFiles()) {
+            if (targetFile.exists()) {
+                throw new DownloadException("Target file exists: " + targetFile.getAbsolutePath());
+            }
+        }
+
+        final Map<URI, File> cachedFiles = fetchingCacheService.getAll(downloadsManifest.getSourceFileUris());
+
+        for (final Map.Entry<URI, File> uriFileEntry : downloadsManifest.getEntries()) {
+
+            final File cachedFile = cachedFiles.get(uriFileEntry.getKey());
+
+            if (cachedFile == null || !cachedFile.exists()) {
+                throw new DownloadException("Failed get file handle from cache: " + uriFileEntry.getKey());
+            }
+
+            try {
+                Files.copy(cachedFile, uriFileEntry.getValue());
```

Comment:
You're probably right.
The cache owns the files, giving out handles to them makes it hard to for example garbage-collect old versions because who knows what is using them.

Reply:
So you're moving this?

Reply:
Yes